File: src/ml/utils/data_splitter.py
"""
Time-series aware data splitting utilities for ML training.

Implements proper time-based splits to prevent lookahead bias
and supports walk-forward validation strategies.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import List, Tuple, Optional
import pandas as pd
import numpy as np
import structlog

logger = structlog.get_logger(__name__)


@dataclass
class DataSplit:
    """Container for train/val/test data splits."""
    train: pd.DataFrame
    val: pd.DataFrame
    test: pd.DataFrame
    
    @property
    def train_size(self) -> int:
        return len(self.train)
    
    @property
    def val_size(self) -> int:
        return len(self.val)
    
    @property
    def test_size(self) -> int:
        return len(self.test)
    
    def summary(self) -> str:
        """Return summary of split sizes."""
        total = self.train_size + self.val_size + self.test_size
        return (
            f"Split summary:\n"
            f"  Train: {self.train_size} ({100*self.train_size/total:.1f}%)\n"
            f"  Val:   {self.val_size} ({100*self.val_size/total:.1f}%)\n"
            f"  Test:  {self.test_size} ({100*self.test_size/total:.1f}%)\n"
            f"  Total: {total}"
        )


class DataSplitter:
    """
    Time-series aware data splitting for ML training.
    
    Ensures proper temporal ordering to prevent lookahead bias.
    Supports both fixed ratio splits and walk-forward validation.
    """
    
    def __init__(
        self,
        train_ratio: float = 0.7,
        val_ratio: float = 0.15,
        test_ratio: float = 0.15,
    ):
        """
        Initialize data splitter with split ratios.
        
        Args:
            train_ratio: Proportion of data for training (default 0.7)
            val_ratio: Proportion of data for validation (default 0.15)
            test_ratio: Proportion of data for testing (default 0.15)
        
        Raises:
            ValueError: If ratios don't sum to 1.0
        """
        if not abs(train_ratio + val_ratio + test_ratio - 1.0) < 1e-6:
            raise ValueError(
                f"Split ratios must sum to 1.0, got {train_ratio + val_ratio + test_ratio}"
            )
        
        if train_ratio <= 0 or val_ratio <= 0 or test_ratio <= 0:
            raise ValueError("All split ratios must be positive")
        
        self.train_ratio = train_ratio
        self.val_ratio = val_ratio
        self.test_ratio = test_ratio
    
    def split(self, df: pd.DataFrame, ensure_sorted: bool = True) -> DataSplit:
        """
        Split dataframe into train/val/test sets using time-based ordering.
        
        Args:
            df: Input dataframe with datetime index or timestamp column
            ensure_sorted: If True, sort by index/timestamp before splitting
        
        Returns:
            DataSplit containing train, val, and test dataframes
        
        Raises:
            ValueError: If dataframe is empty or too small to split
        """
        if df.empty:
            raise ValueError("Cannot split empty dataframe")
        
        if len(df) < 10:
            raise ValueError(
                f"Dataframe too small to split (size={len(df)}). Need at least 10 rows."
            )
        
        # Work with a copy
        data = df.copy()
        
        # Ensure temporal ordering
        if ensure_sorted:
            if isinstance(data.index, pd.DatetimeIndex):
                data = data.sort_index()
            elif 'timestamp' in data.columns:
                data = data.sort_values('timestamp').reset_index(drop=True)
            else:
                logger.warning(
                    "No datetime index or timestamp column found. "
                    "Assuming data is already in temporal order."
                )
        
        # Calculate split indices
        n = len(data)
        train_end = int(n * self.train_ratio)
        val_end = train_end + int(n * self.val_ratio)
        
        # Perform split (time-ordered, no shuffling)
        train_df = data.iloc[:train_end].copy()
        val_df = data.iloc[train_end:val_end].copy()
        test_df = data.iloc[val_end:].copy()
        
        # Validate splits are non-empty
        if len(train_df) == 0 or len(val_df) == 0 or len(test_df) == 0:
            raise ValueError(
                f"Split resulted in empty partition. "
                f"Sizes: train={len(train_df)}, val={len(val_df)}, test={len(test_df)}"
            )
        
        split = DataSplit(train=train_df, val=val_df, test=test_df)
        logger.info(
            "Data split completed",
            train_size=split.train_size,
            val_size=split.val_size,
            test_size=split.test_size,
        )
        
        return split
    
    def walk_forward_splits(
        self,
        df: pd.DataFrame,
        n_splits: int = 5,
        min_train_size: Optional[int] = None,
    ) -> List[Tuple[pd.DataFrame, pd.DataFrame]]:
        """
        Generate walk-forward validation splits for time-series cross-validation.
        
        Each split uses an expanding training window and a fixed-size test window.
        This prevents lookahead bias while maximizing data utilization.
        
        Args:
            df: Input dataframe with datetime index or timestamp column
            n_splits: Number of CV splits to generate (default 5)
            min_train_size: Minimum training set size (default: 30% of data)
        
        Returns:
            List of (train_df, val_df) tuples for each fold
        
        Raises:
            ValueError: If dataframe is too small for the requested splits
        """
        if df.empty:
            raise ValueError("Cannot split empty dataframe")
        
        if n_splits < 2:
            raise ValueError(f"n_splits must be at least 2, got {n_splits}")
        
        # Work with a copy and ensure temporal ordering
        data = df.copy()
        if isinstance(data.index, pd.DatetimeIndex):
            data = data.sort_index()
        elif 'timestamp' in data.columns:
            data = data.sort_values('timestamp').reset_index(drop=True)
        
        n = len(data)
        
        # Set minimum training size
        if min_train_size is None:
            min_train_size = int(n * 0.3)
        
        if min_train_size >= n:
            raise ValueError(
                f"min_train_size ({min_train_size}) must be less than data size ({n})"
            )
        
        # Calculate test window size for each split
        remaining_data = n - min_train_size
        test_size = remaining_data // n_splits
        
        if test_size < 1:
            raise ValueError(
                f"Not enough data for {n_splits} splits. "
                f"Data size={n}, min_train_size={min_train_size}"
            )
        
        # Generate expanding window splits
        splits = []
        for i in range(n_splits):
            train_end = min_train_size + (i * test_size)
            val_end = train_end + test_size
            
            if val_end > n:
                val_end = n
            
            train_df = data.iloc[:train_end].copy()
            val_df = data.iloc[train_end:val_end].copy()
            
            if len(val_df) > 0:  # Only add if validation set is non-empty
                splits.append((train_df, val_df))
        
        logger.info(
            "Walk-forward splits generated",
            n_splits=len(splits),
            min_train_size=min_train_size,
            test_size=test_size,
        )
        
        return splits
    
    @staticmethod
    def verify_no_leakage(
        train_df: pd.DataFrame,
        test_df: pd.DataFrame,
        timestamp_col: Optional[str] = None,
    ) -> bool:
        """
        Verify that no temporal leakage exists between train and test sets.
        
        Args:
            train_df: Training dataframe
            test_df: Test dataframe
            timestamp_col: Name of timestamp column (if not using datetime index)
        
        Returns:
            True if no leakage detected, False otherwise
        """
        if train_df.empty or test_df.empty:
            logger.warning("Cannot verify leakage on empty dataframe")
            return True
        
        # Get timestamps
        if isinstance(train_df.index, pd.DatetimeIndex):
            train_max_time = train_df.index.max()
            test_min_time = test_df.index.min()
        elif timestamp_col and timestamp_col in train_df.columns:
            train_max_time = train_df[timestamp_col].max()
            test_min_time = test_df[timestamp_col].min()
        else:
            logger.warning(
                "No datetime index or timestamp column found. Cannot verify leakage."
            )
            return True
        
        # Check for leakage
        has_leakage = train_max_time >= test_min_time
        
        if has_leakage:
            logger.error(
                "Temporal leakage detected!",
                train_max_time=str(train_max_time),
                test_min_time=str(test_min_time),
            )
        else:
            logger.info(
                "No temporal leakage detected",
                train_max_time=str(train_max_time),
                test_min_time=str(test_min_time),
            )
        
        return not has_leakage

File: cpp-cohida/src/ml/DataSplitter.cpp
#include "ml/DataSplitter.h"
#include "utils/Logger.h"
#include <cmath>
#include <sstream>
#include <stdexcept>

namespace ml {

// --------------------------------------------------------------- DataSplit
std::string DataSplit::summary() const {
  int total = train_size() + val_size() + test_size();
  std::ostringstream oss;
  oss << "Split summary:\n"
      << "  Train: " << train_size() << " (" << (100.0 * train_size() / total)
      << "%)\n"
      << "  Val:   " << val_size() << " (" << (100.0 * val_size() / total)
      << "%)\n"
      << "  Test:  " << test_size() << " (" << (100.0 * test_size() / total)
      << "%)\n"
      << "  Total: " << total;
  return oss.str();
}

// --------------------------------------------------------------- DataSplitter
DataSplitter::DataSplitter(double train_ratio, double val_ratio,
                           double test_ratio)
    : train_ratio_(train_ratio), val_ratio_(val_ratio),
      test_ratio_(test_ratio) {
  if (std::abs(train_ratio + val_ratio + test_ratio - 1.0) > 1e-6)
    throw std::invalid_argument("Split ratios must sum to 1.0");
  if (train_ratio <= 0 || val_ratio <= 0 || test_ratio <= 0)
    throw std::invalid_argument("All split ratios must be positive");
}

DataSplit DataSplitter::split(const Eigen::MatrixXd &X,
                              const Eigen::VectorXd &y) const {
  const int n = static_cast<int>(X.rows());
  if (n == 0)
    throw std::invalid_argument("Cannot split empty data");
  if (n < 10)
    throw std::invalid_argument(
        "Data too small to split (need at least 10 rows)");
  if (X.rows() != y.size())
    throw std::invalid_argument("X and y must have same number of rows");

  int train_end = static_cast<int>(n * train_ratio_);
  int val_end = train_end + static_cast<int>(n * val_ratio_);

  // Ensure non-empty partitions
  if (train_end < 1)
    train_end = 1;
  if (val_end <= train_end)
    val_end = train_end + 1;
  if (val_end >= n)
    val_end = n - 1;

  DataSplit result;
  result.train = X.middleRows(0, train_end);
  result.val = X.middleRows(train_end, val_end - train_end);
  result.test = X.middleRows(val_end, n - val_end);

  result.train_target = y.segment(0, train_end);
  result.val_target = y.segment(train_end, val_end - train_end);
  result.test_target = y.segment(val_end, n - val_end);

  LOG_INFO("Data split: train={}, val={}, test={}", result.train_size(),
           result.val_size(), result.test_size());

  return result;
}

std::vector<DataSplitter::WalkForwardFold>
DataSplitter::walk_forward_splits(const Eigen::MatrixXd &X,
                                  const Eigen::VectorXd &y, int n_splits,
                                  std::optional<int> min_train_size) const {

  const int n = static_cast<int>(X.rows());
  if (n == 0)
    throw std::invalid_argument("Cannot split empty data");
  if (n_splits < 2)
    throw std::invalid_argument("n_splits must be at least 2");

  int min_train = min_train_size.value_or(static_cast<int>(n * 0.3));
  if (min_train >= n)
    throw std::invalid_argument("min_train_size must be less than data size");

  int remaining = n - min_train;
  int test_size = remaining / n_splits;
  if (test_size < 1)
    throw std::invalid_argument("Not enough data for requested splits");

  std::vector<WalkForwardFold> folds;
  for (int i = 0; i < n_splits; ++i) {
    int train_end = min_train + i * test_size;
    int val_end = std::min(train_end + test_size, n);

    if (val_end <= train_end)
      break;

    WalkForwardFold fold;
    fold.train_X = X.middleRows(0, train_end);
    fold.train_y = y.segment(0, train_end);
    fold.val_X = X.middleRows(train_end, val_end - train_end);
    fold.val_y = y.segment(train_end, val_end - train_end);
    folds.push_back(std::move(fold));
  }

  LOG_INFO("Walk-forward splits generated: {} folds, min_train={}",
           static_cast<int>(folds.size()), min_train);
  return folds;
}

bool DataSplitter::verify_no_leakage(
    const std::vector<std::chrono::system_clock::time_point> &timestamps,
    int train_end, int test_start) {

  if (timestamps.empty() || train_end < 0 || test_start < 0)
    return true;
  if (train_end >= static_cast<int>(timestamps.size()) ||
      test_start >= static_cast<int>(timestamps.size()))
    return true;

  bool has_leakage = timestamps[static_cast<size_t>(train_end)] >=
                     timestamps[static_cast<size_t>(test_start)];
  if (has_leakage) {
    LOG_ERROR("Temporal leakage detected!");
  }
  return !has_leakage;
}

} // namespace ml
